### ISSUE TITLE
Focus loss when “Clear” button is used in Styles > Color

### ISSUE THREAD

**Issue by stokesman** [WordPress/gutenberg#60614]
### Description

In the Styles > Colors panel (or the Styles > Blocks > _{block}_ > _{block variation}_ panel) the “Clear” button is removed when the color is cleared. This results in a focus loss.

Color editing on a individual block does not suffer from this.

Apart from the specific issue this raises the question of whether the `ColorPalette` component could be enhanced to make this problematic scenario less likely. Hiding the clear button when the color is the default seems like it was done with good intent. It’s an indicator that the default is selected so there is nothing to clear. However that would be better done by disabling the button yet it appears `ColorPalette` does not support such usage. Perhaps it should.

### Step-by-step reproduction instructions

1. In the Site Editor
2. Go to Styles > Colors
3. Change a color to something other than the default
4. Use the “Clear” button
5. Notice focus is lost (palette dropdown doesn’t close as expected or if you are using the keyboard you're nowhere near where you were)

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/9000376/57134210-c503-4c23-a4de-b27cf96dccb1

### Environment info

- Gutenberg Trunk

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by Rishit30G** [comment]
Thanks for sharing the issue, 

I can confirm that the issue still persists, sharing the screencast for the same: 

Gutenberg: Version 20.2.0
WordPress: 6.7.1
Theme: TT4

https://github.com/user-attachments/assets/019b6e30-ab88-4ed4-a500-ef1f666fa7dc

**Comment by Rishit30G** [comment]
Hey @stokesman, 
Sharing a screencast of how I have implemented the solution that you have shared, please share your thoughts on this one 🙇🏻 : 

https://github.com/user-attachments/assets/414786f0-c935-4756-83ee-fef663eed9f8

**Comment by Mamaduka** [comment]
Making the "Clear" button always visible could be considered as a breaking change, even when it's done with good intentions.

Should we try to solve this in the `ColorPanelTab` component?

https://github.com/WordPress/gutenberg/blob/177743059a87db2ba5f11f83dc8964e543bf3a03/packages/block-editor/src/components/global-styles/color-panel.js#L194

**Comment by Rishit30G** [comment]
Thanks for the suggestion @Mamaduka 👍🏻 

So to avoid making the "Clear" button visible at all times, we can add an 'or' condition in the `clearable` like this, allowing multiple conditions to address the issue more effectively.

However we would need to disable the clear button when the userValue is undefined, that we should be good 
```jsx 
clearable={ inheritedValue === userValue || userValue === undefined }
```

**Comment by stokesman** [comment]
If the ideal resolution is for this to work just like when editing color directly on a block (that is, to have the “Clear” button always present and only disabled when the user has picked a color) it’s does not seem possible without some extension of the `ColorPalette` component. That’s because whether the button is enabled isn’t configurable outside the component and instead is based on whether its `value` prop is falsy.

In @Rishit30G’s latest suggestion, the “Clear“ button would always be present yet never disabled (in the Styles panel). That’s a less grave issue than the focus loss so it’d seem a step worth taking. Though it could be more sensibly done just by removing the `clearable` prop—it defaults to `true`

<details><summary>An alternative would be to always hide the “Clear” button in the Global Styles panels. It doesn’t disable the resetting of the value—there are two ways to do that besides the button. It would prevent the focus problem and avoid there sometimes being an enabled “Clear” button that doesn’t do anything. Disclosed is a diff of how that might be done.</summary>

```diff
diff --git a/packages/block-editor/src/components/global-styles/color-panel.js b/packages/block-editor/src/components/global-styles/color-panel.js
index 5d5c02d179..53d203b66c 100644
--- a/packages/block-editor/src/components/global-styles/color-panel.js
+++ b/packages/block-editor/src/components/global-styles/color-panel.js
@@ -180,6 +180,7 @@ function ColorPanelTab( {
 	userValue,
 	setValue,
 	colorGradientControlSettings,
+	isBlockScope,
 } ) {
 	return (
 		<ColorGradientControl
@@ -191,7 +192,7 @@ function ColorPanelTab( {
 			gradientValue={ isGradient ? inheritedValue : undefined }
 			onColorChange={ isGradient ? undefined : setValue }
 			onGradientChange={ isGradient ? setValue : undefined }
-			clearable={ inheritedValue === userValue }
+			clearable={ isBlockScope && inheritedValue === userValue }
 			headingLevel={ 3 }
 		/>
 	);
@@ -210,6 +211,7 @@ function ColorPanelDropdown( {
 	const currentTab = tabs.find( ( tab ) => tab.userValue !== undefined );
 	const { key: firstTabKey, ...firstTab } = tabs[ 0 ] ?? {};
 	const colorGradientDropdownButtonRef = useRef( undefined );
+	const isBlockScope = !! panelId;
 	return (
 		<ToolsPanelItem
 			className="block-editor-tools-panel-color-gradient-settings__item"
@@ -271,6 +273,7 @@ function ColorPanelDropdown( {
 									colorGradientControlSettings={
 										colorGradientControlSettings
 									}
+									isBlockScope={ isBlockScope }
 								/>
 							) }
 							{ tabs.length > 1 && (
@@ -301,6 +304,9 @@ function ColorPanelDropdown( {
 													colorGradientControlSettings={
 														colorGradientControlSettings
 													}
+													isBlockScope={
+														isBlockScope
+													}
 												/>
 											</Tabs.TabPanel>
 										);

```
</details>

I’m leaning a bit toward the latter as a better solution for now. That’s because in some cases that the “Clear” button can have an effect it’s more akin to a “Reset” and that seems like a bit of an issue as well.

**Comment by Rishit30G** [comment]
Appreciate your thoughts and solution @stokesman 👏🏻 

> Though it could be more sensibly done just by removing the clearable prop—it defaults to true

I could have gone with this approach as well, but I wanted the `clearable` boolean value to be passed in the [`disable` prop ](https://github.com/WordPress/gutenberg/pull/69097/files#:~:text=disabled%3D%7B%20!%20value%20%7C%7C%20!%20clearable%20%7D)so that I can show the "Clear" button as disabled when the colour hasn't changed


> I’m leaning a bit toward the latter as a better solution for now.

I tested out this solution on local, it looks good to me, I can update my PR accordingly if we plan to go ahead with this one 👍🏻 

> That’s because in some cases that the “Clear” button can have an effect it’s more akin to a “Reset” and that seems like a bit of an issue as well.

Makes sense. 👍🏻 

Having a Clear/Reset button is helpful. If a user selects a new color from the palette and wants to revert to the original color, finding the exact color can be time-consuming. They might have to try multiple colors to get back to the original one. A Clear/Reset button eliminates this hassle, allowing users to revert in just one step.

This is just my perspective feel free to share any constructive counterpoints! 😄

Thanks

**Comment by stokesman** [comment]
I wrote my [commentary](https://github.com/WordPress/gutenberg/issues/60614#:~:text=In%20%40Rishit30G%E2%80%99s%20latest%20suggestion) regarding your most recent suggestion https://github.com/WordPress/gutenberg/issues/60614#issuecomment-2647073353 under the (apparently mistaken) impression that it was in the context of trunk and not along with the change in your PR’s branch. I think we are not quite on the same page yet. Let me go back to part of your comment.

> So to avoid making the "Clear" button visible at all times

In your PR, it’s currently unavoidable. The button is unconditionally rendered.

https://github.com/WordPress/gutenberg/blob/208c84246e9c38a236607ca8f8fca1ce83318d9c/packages/components/src/color-palette/index.tsx#L244-L252

That’s not a problem in the context of Gutenberg but for third-parties that may be using the `ColorPalette` component. The `clearable` prop no longer does what it used to—controlling the presence of the clear button. It’s against the policy on backwards compatibility to change that. This is what George was getting at https://github.com/WordPress/gutenberg/issues/60614#issuecomment-2644721684.

--- 
> Having a Clear/Reset button is helpful.

I agree and it’s consistent with the experience when setting colors on a block. A user might miss the fact that clicking the already selected swatch is a way to clear it. Or they might not be familiar with the "Reset" button within the ellipsis menu of the panel. So I can appreciate it may be better to keep the consistency despite the fact that clearing the color is possible without it.

--- 

In the interest of trying the smallest step that can be considered worthwhile, I think we could just remove the `clearable` prop. It would fix the focus loss. Then we’d be left with the issue that the button is always enabled even if the color is the initial one. It’s a tradeoff but a favorable one IMO. If you agree and want to revamp your PR to that effect I think it would have decent chance of landing.

**Comment by Rishit30G** [comment]
> I wrote my https://github.com/WordPress/gutenberg/issues/60614#:~:text=In%20%40Rishit30G%E2%80%99s%20latest%20suggestion regarding your most recent suggestion https://github.com/WordPress/gutenberg/issues/60614#issuecomment-2647073353 under the (apparently mistaken) impression that it was in the context of trunk and not along with the change in your PR’s branch.

Sorry for the overlook. 🙇🏻 

> It’s against the policy on backwards compatibility to change that. This is what George was getting at https://github.com/WordPress/gutenberg/issues/60614#issuecomment-2644721684.

Understood now, thanks! 👍🏻 

> If you agree and want to revamp your PR to that effect I think it would have decent chance of landing.

I agree with the statements shared above, have revamped my PR accordingly

**Comment by t-hamano** [comment]
> If the ideal resolution is for this to work just like when editing color directly on a block (that is, to have the “Clear” button always present and only disabled when the user has picked a color) it’s does not seem possible without some extension of the `ColorPalette` component. That’s because whether the button is enabled isn’t configurable outside the component and instead is based on whether its `value` prop is falsy.

I tried the following code, passing `undefied` to `colorValue` to disable the clear button if there is no user value. This seems to work fine, but what do you think?

```diff
diff --git a/packages/block-editor/src/components/global-styles/color-panel.js b/packages/block-editor/src/components/global-styles/color-panel.js
index 5d5c02d179..7bd5f5c5af 100644
--- a/packages/block-editor/src/components/global-styles/color-panel.js
+++ b/packages/block-editor/src/components/global-styles/color-panel.js
@@ -187,11 +187,12 @@ function ColorPanelTab( {
                        showTitle={ false }
                        enableAlpha
                        __experimentalIsRenderedInSidebar
-                       colorValue={ isGradient ? undefined : inheritedValue }
+                       colorValue={
+                               isGradient || ! userValue ? undefined : inheritedValue
+                       }
                        gradientValue={ isGradient ? inheritedValue : undefined }
                        onColorChange={ isGradient ? undefined : setValue }
                        onGradientChange={ isGradient ? setValue : undefined }
-                       clearable={ inheritedValue === userValue }
                        headingLevel={ 3 }
                />
        );
```

https://github.com/user-attachments/assets/2228bac3-83ec-415b-b380-ee7caa7060c8

**Comment by Rishit30G** [comment]
After going few components deep, I could understand that the colorValue and clearable work together
So this approach seems to be good, I have updated my PR accordingly! 
Thanks @t-hamano! 🚀 

Would appreciate a quick PR review as well 👍🏻

**Comment by t-hamano** [comment]
> I tried the following code, passing `undefied` to `colorValue` to disable the clear button if there is no user value. This seems to work fine, but what do you think?

Sorry, this approach was incorrect. `The inherited value is not reflected in the palette:

![Image](https://github.com/user-attachments/assets/f2f1e07c-3945-40cd-abc5-91a4ed4fe669)

**Comment by Rishit30G** [comment]
> The inherited value is not reflected in the palette:

Oh yes, but we can only have one value being passed in the `colorValue` prop, which is either `undefined` or `inheritedValue`
As per the design we need both the values to be passed, `undefined` to make it disabled and `inheritedValue` to get the color in rectangular palette color 

PS: In the post, if I use the panel to add and clear color, the behaviour is same as this https://github.com/WordPress/gutenberg/issues/60614#issuecomment-2788286073 🤔 

<img width="400" height="250" alt="Image" src="https://github.com/user-attachments/assets/dfd380b7-9cab-4010-ad53-e61d9ccdc4ad" />

<img width="400" height="250" alt="Image" src="https://github.com/user-attachments/assets/42211dc8-1e89-4033-a417-f3627a980cd3" />
